### DISCUSSION TITLE
Enhancement: Add setting to change behavior when type information isn't available for rules that require it

### DISCUSSION THREAD

**Discussion by RebeccaStevens** [typescript-eslint/typescript-eslint#6519] [Technical Discussions]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).
  - It's a little niche mut this can't really be done anywhere else.

### Template

None of the templates really fit for this so I hope you don't mind me going off template.

### Description

Currently when you try to use a rule that requires type information, but it isn't available, you get this error message:
`Error: You have used a rule which requires parserServices to be generated. You must therefore provide a value for the "parserOptions.project" property for @typescript-eslint/parser.`

I'd like to propose adding a [shared setting](https://eslint.org/docs/latest/use/configure/configuration-files#adding-shared-settings) that can change this. The options I propose are `error` (the current behavior), `warn` and `ignore`.
In the case of `warn` and `ignore`, the rule will simply just not run in the case where type info isn't available rather than erroring out the whole eslint task.

A similar setting could also be added to specify the behavior for rules that use type info as an enhancer but don't out right require it.

Having this plugin add these settings would allow other plugins to leverage the same settings to use themselves. It would be nice to have relevant utility functions available for other plugins to use to make it really easy from them to implement this behavior.

### Use Cases

A simple use case would be for [eslint-plugin-markdown](https://www.npmjs.com/package/eslint-plugin-markdown). As that processor can't leverage type info, it would be nice to just ignore rules that require type info rather than having to explicitly disable each one.

**Comment by JoshuaKGoldberg** [comment]
This is an interesting idea! Let's move to a discussion so we get nice threaded comments.

**Comment by JoshuaKGoldberg** [comment]
> A simple use case would be for [eslint-plugin-markdown](https://www.npmjs.com/package/eslint-plugin-markdown). As that processor can't leverage type info, it would be nice to just ignore rules that require type info rather than having to explicitly disable each one.

Just to clarify - is this the same as https://github.com/JoshuaKGoldberg/template-typescript-node-package/issues/208? If not, do you have an example of the situation you're trying to avoid?

**Reply by RebeccaStevens** [reply]
Mostly what I was referring to is not having to do this: https://github.com/eslint-functional/eslint-plugin-functional/blob/main/.eslintrc.json#L98-L149

**Reply by JoshuaKGoldberg** [reply]
Oh fun! In that case I'll defer to https://github.com/typescript-eslint/typescript-eslint/discussions/6519#discussioncomment-5084492 as a suggested solution.

**Comment by bradzacher** [comment]
I don't think this is the best solution because that error is there to expose real configuration problems in your codebase.

We unfortunately cannot "warn" via logging because there are many usecases where you don't have access to the console logs (i.e. IDE) and also additional console logs break usecases that require pure JSON output from the lint process (some IDE integrations or custom lint scripts).

Silently ignoring these cases is bad because it can easily create ***false negatives*** - cases where we should report, but due to a misconfiguration we don't report. False negatives are a ***massive*** problem because as a user you will see no lint reports and think your code is safe and sound, only to find out that it's not - usually later when your code breaks in production.
This irrevocably harms trust in the linting ecosystem - which is obviously very bad!

**Reply by RebeccaStevens** [reply]
Do you have a recommended way for rule that use type info as an enhancement rather than a requirement, to report config errors?

**Reply by bradzacher** [reply]
If it's an optional enhancement I would probably put it behind a flag and hard-error when that flag is on.

That way someone needs to explicitly opt-in and they get errors if something is wrong with their config.

I know in an ideal world you would want to just progressively enhance the rule, but currently at least there's just a lot of avenues for misconfiguration - so you want to avoid accidental false negatives.

**Comment by bradzacher** [comment]
In our upcoming v6 release we're adding a config which will disable type-information and our type-aware lint rules:
https://github.com/typescript-eslint/typescript-eslint/blob/217c710d99445994b9c8db7b9bee9b9cc63bc4cb/.eslintrc.js#L207-L209
Which can be used to quickly turn off type-aware linting for files that don't have type information.

There should be a way to select processor files in `overrides` config - I think the pattern `**/*.md/*.ts` will probably do it? I'm not 100% sure though.

**Reply by JoshuaKGoldberg** [reply]
That solution should work for https://github.com/JoshuaKGoldberg/template-typescript-node-package/issues/208 in general. Though it is a pity that it assumes there will be no folder named `*.md` - which can theoretically happen.
